subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/version/CommandLineVersionLocator.java

```diff
@@ -0,0 +1,169 @@
+    private static final String VSWHERE_EXE = "vswhere.exe";
+    private static final String INSTALLATION_PATH_KEY = "installationPath";
+    private static final String INSTALLATION_VERSION_KEY = "installationVersion";
+
+    public CommandLineVersionLocator(ExecActionFactory execActionFactory, WindowsRegistry windowsRegistry, OperatingSystem os, VisualCppMetadataProvider visualCppMetadataProvider) {
+        this.execActionFactory = execActionFactory;
+        this.windowsRegistry = windowsRegistry;
+        this.os = os;
+        this.visualCppMetadataProvider = visualCppMetadataProvider;
+    }
+
+    @Override
+    protected List<VisualStudioMetadata> locateInstalls() {
+        List<VisualStudioMetadata> installs = Lists.newArrayList();
+
+        File vswhereBinary = findVswhereBinary();
+        if (vswhereBinary != null) {
+            List<String> args = Lists.newArrayList("-legacy", "-format", "json");
+            String json = getVswhereOutput(vswhereBinary, args);
+            installs.addAll(parseJson(json));
+        }
+
+        return installs;
+    }
+
+    private File findVswhereBinary() {
+        for (String programFilesKey : PROGRAM_FILES_KEYS) {
+            File programFilesDir;
+            try {
+                programFilesDir = new File(windowsRegistry.getStringValue(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, REGISTRY_PATH_WINDOWS, programFilesKey));
+            } catch (MissingRegistryEntryException e) {
+                // We'll get this when we try to look up "ProgramFilesDir (x86)" on a 32-bit OS
+                continue;
+            }
+
+            File candidate = new File(programFilesDir, VISUAL_STUDIO_INSTALLER + "/" + VSWHERE_EXE);
+            if (candidate.exists() && candidate.isFile()) {
+                return candidate;
+            }
+        }
+
+        return os.findInPath(VSWHERE_EXE);
+    }
+
+    private String getVswhereOutput(File vswhereBinary, List<String> args) {
+        ExecAction exec = execActionFactory.newExecAction();
+        exec.args(args);
+        exec.executable(vswhereBinary.getAbsolutePath());
+        exec.setWorkingDir(vswhereBinary.getParentFile());
+
+        StreamByteBuffer buffer = new StreamByteBuffer();
+        exec.setStandardOutput(buffer.getOutputStream());
+        exec.setErrorOutput(NullOutputStream.INSTANCE);
+        exec.setIgnoreExitValue(true);
+        ExecResult result = exec.execute();
+
+        int exitValue = result.getExitValue();
+        if (exitValue == 0) {
+            return buffer.readAsString();
+        } else {
+            LOGGER.debug("vswhere.exe returned a non-zero exit value ({}) - ignoring", result.getExitValue());
+            return null;
+        }
+    }
+
+    private List<VisualStudioMetadata> parseJson(String json) {
+        List<VisualStudioMetadata> installs = Lists.newArrayList();
+        JsonReader reader = new JsonReader(new StringReader(json));
```

Comment:
Do we need to worry about the encoding of this file? I think it's unicode based on https://github.com/Microsoft/vswhere/wiki/Encoding

Reply:
Good thought.  I've changed the encoding to explicitly UTF-8.